tensorflow/core/transforms/remapper/pass.cc

```diff
@@ -306,6 +312,241 @@ class BasePatternActivationRewriter : public BasePatternRewriter {
+      // cuDNN only supports NHWC data layout.
+      if (data_format != "NHWC") return false;
+
+      // Data type must be Float16.
+      if (!dtype_T.isa<Float16Type>()) return false;
+
+      // Channel dimension must be a multiple of 4.
+      auto fbn_input0_shape =
+          fused_batch_norm_op->getOperand(0).getType().cast<ShapedType>();
+      auto fbn_input0_shape_dims = fbn_input0_shape.getShape();
+
+      const bool valid_channel_dim = (fbn_input0_shape.getRank() == 4) &&
+                                     (fbn_input0_shape_dims[3] % 4 == 0);
+
+      if (!valid_channel_dim) return false;
+
+      // cuDNN must support CUDNN_BATCHNORM_SPATIAL_PERSISTENT mode.
+      if (!BatchnormSpatialPersistentEnabled()) return false;
+    }
+
+    // FusedBatchNormV2 and V3 have an extra type parameter.
+    if (fused_batch_norm_op->getName().getStringRef().str() !=
+        "tfg.FusedBatchNorm") {
+      auto attr = fused_batch_norm_op->getAttrOfType<TypeAttr>("U");
+      if (attr && !attr.getValue().isa<Float32Type>()) {
+        return false;
+      }
+    }
+
+    // Check that only one node consumes the 0-th output of a FusedBatchNorm.
+    if (this->helper_.HasControlOperandsOrResultUsers(fused_batch_norm_op) ||
+        !this->helper_.HasAtMostOneUserOfResult0(fused_batch_norm_op))
+      return false;
+
+    return true;
+  }
+
+  bool matchPattern(Operation *op, Pattern &pattern) const {
+    TFOp activation_tfg_wrapper(op);
+    // Not allowing control flow on Relu
+    if (helper_.HasControlOperandsOrResultUsers(op)) return false;
+    if (activation_tfg_wrapper.getNonControlOperands().empty()) return false;
+
+    Operation *activation_input_op = op->getOperand(0).getDefiningOp();
+    if (is_valid_batch_norm(activation_input_op)) {
+      pattern.fused_batch_norm = activation_input_op;
+      pattern.activation = op;
+      pattern.side_input = nullptr;
+      return true;
+    }
+
+    // Input to a Relu can be an Add node with FusedBatchNorm as one of the
+    // inputs
+    if (this->helper_.getDialect()->IsAdd(activation_input_op)) {
+      // Currently no CPU implementation for "FusedBatchNorm + SideInput +
+      // <Activation>"
+      if (this->helper_.isOneDNNEnabled() &&
+          !util::OpHasDevice(op, tensorflow::DEVICE_GPU))
+        return false;
+
+      // Check that only Relu node consumes the output of an Add node.
+      if (helper_.HasControlOperandsOrResultUsers(activation_input_op) ||
+          !helper_.HasAtMostOneUserOfResult0(activation_input_op))
+        return false;
+
+      if (activation_input_op->getOperands().size() < 2 &&
+          TFOp(activation_input_op).getNonControlOperands().size() < 2)
+        return false;
+
+      // Add node supports broadcasting, FusedBatchNormEx does not.
+      // Check for symbolic shape equivalence
+      auto add_input0_op = activation_input_op->getOperand(0).getDefiningOp();
+      auto add_input1_op = activation_input_op->getOperand(1).getDefiningOp();
```

Comment:
Same as above. Not sure if we need to check nullity of definingOp here

Reply:
Added additional check for this.